### PR TITLE
Issue 3638: Ensure the AsyncSegmentInputStreamImpl handles ConnectionFailedException during sendAsync.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
- Ensure AsyncSegmentInputStreamImpl closes the connection incase a `ConnectionFailedException` is observed during `io.pravega.client.netty.impl.ClientConnection#sendAsync()`

**Purpose of the change**  
Related to issue #3638

**What the code does**  
AsycnSegmentInputStreamImpl invokes `ClientConnection.sendAsync(...)` to send commands like `WireCommand.ReadSegment`. This method can throw a `ConnectionFailedException` incase the underlying Netty Channel was set to null due to `channelUnregistered()` callback by netty. This can cause the AsyncSegmentInputStream to retry until the replyProcessor.connectionDropped() is invoked by `FlowHandler`.

**How to verify it**  
All the existing and newly added tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java [hunk 1]
```diff
@@ -181,11 +180,17 @@ public boolean isClosed() {
...
                         if (ex != null) {
-                            log.warn("Exception while establishing connection with Pravega " +
-                                    "node", ex);
+                            log.warn("Exception while establishing connection with Pravega node", ex);
                             closeConnection(new ConnectionFailedException(ex));
                         }
-                    }).thenCompose(c -> sendRequestOverConnection(request, c));
+                    }).thenCompose(c -> sendRequestOverConnection(request, c)
+                            .whenComplete((reply, ex) -> {
+                                if (ex instanceof ConnectionFailedException) {
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: Shouldn't you unwrap this exception?

Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: Consider using `Futures.exceptionallyExpecting` which will handle all of this and make your code more compact.

**Comment by shrids** [comment]
Closing this PR, since we are not observing with the latest client code.

**Comment by shrids** [comment]
Reopening this as this error was observed again.

File: client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: The future returned  `sendRequestOverConnecton` is used by the overall retry logic. Hence 
```
Futures.exceptionallyExpecting(future, t -> t instanceof ConnectionFailedException, value)
``` 
cannot be used. Do let me know if there is a better way to compact the code.
(the only other option is to invoke `closeConnection(...)` as part of the predicate)

Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java]: Ok. I missed the fact that this is not really returning a value but executing something.

**Pull request merged**